### PR TITLE
fix asyncCall multiTransfer

### PR DESCRIPTION
**Pull request by sasurobert**
fixing nft and multi transfer on asyncCall and asyncCallBack

### CODE REVIEW

File: arwen/host/asyncCall.go [hunk 1]
```diff
@@ -117,12 +135,14 @@ func (host *vmHost) determineAsyncCallExecutionMode(asyncCallInfo *arwen.AsyncCa
 		return arwen.AsyncUnknown, err
 	}
 
-	sameShard := host.AreInSameShard(runtime.GetSCAddress(), asyncCallInfo.Destination)
+	actualDestination := host.determineDestinationForAsyncCall(asyncCallInfo.Destination, functionName, args)
```
Comment (bogdan-rosianu) [inline arwen/host/asyncCall.go:138]: `determineDestinationForAsyncCall` parses esdt transfers but the check for `isESDTTransfer` is below this, so the `determineDestinationForAsyncCall` can be called without the action being an esdt transfer. Is it ok or should we add a check for `isESDTTransfer` inside the function?

File: arwen/host/asyncCall.go [hunk 2]
```diff
@@ -117,12 +135,14 @@ func (host *vmHost) determineAsyncCallExecutionMode(asyncCallInfo *arwen.AsyncCa
...
+
+	sameShard := host.AreInSameShard(runtime.GetSCAddress(), actualDestination)
 	if host.IsBuiltinFunctionName(functionName) {
 		if sameShard {
-			isESDTTransfer, _, _ := host.isESDTTransferOnReturnDataFromFunctionAndArgs(runtime.GetSCAddress(), asyncCallInfo.Destination, functionName, args)
+			isESDTTransfer, _, _ := host.isESDTTransferOnReturnDataFromFunctionAndArgs(runtime.GetSCAddress(), actualDestination, functionName, args)
 			if isESDTTransfer && runtime.GetVMInput().CallType == vm.AsynchronousCall &&
-				bytes.Equal(runtime.GetVMInput().CallerAddr, asyncCallInfo.Destination) {
+				bytes.Equal(runtime.GetVMInput().CallerAddr, actualDestination) {
```
Comment (bogdan-rosianu) [inline arwen/host/asyncCall.go:145]: can you please add unit tests?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: arwen/host/asyncCall.go (hunk 1 above)
Comment (sasurobert) [reply arwen/host/asyncCall.go:138]: no need to add a check inside. as the parseESDTTransfer will have the check.

File: arwen/host/asyncCall.go (hunk 2 above)
Comment (sasurobert) [reply arwen/host/asyncCall.go:145]: tests already exists in mandos json files. I corrected some. And made some to run as expected.

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by matei-p — DISMISSED** [review]

File: arwen/contexts/managedType.go [hunk 3]
```diff
@@ -126,6 +127,7 @@ func (context *managedTypesContext) PopDiscard() {
 // ClearStateStack initializes the state stack
 func (context *managedTypesContext) ClearStateStack() {
 	context.managedTypesStack = make([]managedTypesState, 0)
+	context.randomnessGenerator = nil
```
Comment (iulianpascalau) [inline arwen/contexts/managedType.go:130]: :+1:

**Review by andrei-diaconescu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**